test/unit/util/define-dynamic-block.test.js

```diff
@@ -0,0 +1,202 @@
+        text: 'hat'
+    }
+};
+
+// similar to goog.mixin from the Closure library
+const mixin = function (target, source) {
+    for (const x in source) {
+        target[x] = source[x];
+    }
+};
+
+class MockBlock {
+    // mocks the Blockly.Block constructor
+    static create (blockInfo, extendedOpcode) {
+        // mock the Blockly.Block constructor
+        const block = new MockBlock();
+        const prototype = defineDynamicBlock(MockScratchBlocks, categoryInfo, blockInfo, extendedOpcode);
+        mixin(block, prototype);
+        block.init();
+
+        // bootstrap the mutation<->DOM cycle
+        block.blockInfoText = JSON.stringify(blockInfo);
+        const xmlElement = block.mutationToDom();
+
+        // parse blockInfo from XML to fill dynamic properties
+        block.domToMutation(xmlElement);
+
+        return block;
+    }
+
+    jsonInit (json) {
+        this.result = Object.assign({}, json);
+    }
+    interpolate_ () {
+        // TODO: add tests for this?
+    }
+    setCheckboxInFlyout (isEnabled) {
+        this.result.checkboxInFlyout_ = isEnabled;
+    }
+    setOutput (isEnabled) {
+        this.result.outputConnection = isEnabled; // Blockly calls `makeConnection_` here
+    }
+    setOutputShape (outputShape) {
+        this.result.outputShape_ = outputShape;
+    }
+    setNextStatement (isEnabled) {
+        this.result.nextConnection = isEnabled; // Blockly calls `makeConnection_` here
+    }
+    setPreviousStatement (isEnabled) {
+        this.result.previousConnection = isEnabled; // Blockly calls `makeConnection_` here
+    }
+}
+
+describe('defineDynamicBlock', () => {
+    test('is a function', () => {
+        expect(typeof defineDynamicBlock).toBe('function');
+    });
+    test('can define a command block with an icon', () => {
+        const extendedOpcode = 'test.commandWithIcon';
+        const block = MockBlock.create(testBlockInfo.commandWithIcon, extendedOpcode);
+        expect(block.result).toEqual({
+            category: categoryInfo.name,
+            colour: categoryInfo.color1,
+            colourSecondary: categoryInfo.color2,
+            colourTertiary: categoryInfo.color3,
+            extensions: ['scratch_extension'],
+            inputsInline: true,
+            nextConnection: true,
+            outputShape_: MockScratchBlocks.OUTPUT_SHAPE_SQUARE,
+            previousConnection: true,
+            type: extendedOpcode
+        });
+    });
+    test('can define a command block without an icon', () => {
+        const extendedOpcode = 'test.commandWithoutIcon';
+        const block = MockBlock.create(testBlockInfo.commandWithoutIcon, extendedOpcode);
+        expect(block.result).toEqual({
+            category: categoryInfo.name,
+            colour: categoryInfo.color1,
+            colourSecondary: categoryInfo.color2,
+            colourTertiary: categoryInfo.color3,
+            // extensions: undefined, // no icon means no extension
```

Comment:
Why is this commented out?

Reply:
Or maybe we can add a separate test for the fact that `block.result.extensions === undefined`?

Reply:
Same questions below for undefined properties.

Reply:
Our linting rules forbid the use of `undefined`. Since these tests use `expect(...).toEqual({...})` they would fail if the property is defined. The comments listing these properties as undefined are just meant to emphasize that they're expected to be undefined instead of being don't-care values.

Reply:
Ah, makes sense!